### PR TITLE
Try to evaluate forward refs after model created

### PR DESCRIPTION
**Pull request by uriyyo**
## Change Summary
Try to evaluate forward refs after model created.

So it will not be required to use `update_forward_refs` in the case of self-referencing model:
```python
from __future__ import annotations

from pydantic import BaseModel
from typing import Optional


class Foo(BaseModel):
    foo: Optional[Foo]


assert Foo.__fields__['foo'].type_ is Foo
```

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)

### CODE REVIEW

File: pydantic/main.py [hunk 1]
```diff
@@ -815,6 +817,24 @@ def _get_value(
...
 
+    @classmethod
+    def try_update_forward_refs(cls, **localns: Any) -> None:
+        """
+        Same as update_forward_refs but will not raise exception
+        when forward references are not defined.
+        """
+        if cls.__module__ in sys.modules:
+            globalns = sys.modules[cls.__module__].__dict__.copy()
+        else:
```
Comment (PrettyWood) [inline pydantic/main.py:828]: when is it the case?

**Review by PrettyWood — COMMENTED** [review]
Can you please rebase/merge with master to trigger the new CI?

File: pydantic/main.py (hunk 1 above)
Comment (uriyyo) [reply pydantic/main.py:828]: Without it, some of the tests are failing.

**Comment by samuelcolvin** [comment]
you have conflicts and the wrong change.

Please update.

File: pydantic/main.py [hunk 2]
```diff
@@ -815,6 +817,24 @@ def _get_value(
         else:
             return v
 
+    @classmethod
+    def try_update_forward_refs(cls, **localns: Any) -> None:
```
Comment (samuelcolvin) [inline pydantic/main.py:821]: surely this takes no arguments, it also definitely needs a private name
```suggestion
    def __try_update_forward_refs__(cls) -> None:
```

File: pydantic/main.py [hunk 3]
```diff
@@ -815,6 +817,24 @@ def _get_value(
...
+        """
+        Same as update_forward_refs but will not raise exception
+        when forward references are not defined.
+        """
+        if cls.__module__ in sys.modules:
+            globalns = sys.modules[cls.__module__].__dict__.copy()
+        else:
+            globalns = {}
```
Comment (samuelcolvin) [inline pydantic/main.py:829]:
```suggestion
        module = sys.modules.get(cls.__module__, {'__dict__': {}})
        if module is not None:
            globalns = module.__dict__.copy()
        else:
            globalns = {}
```

Comment (samuelcolvin) [reply pydantic/main.py:829]: maybe this should be duplicated (method in `typing.py` or `utils.py`) and used both here and in `update_forward_refs`?

**Review by samuelcolvin — COMMENTED** [review]
looking good, but this also needs a note in the docs.

File: pydantic/main.py (hunk 3 above)
Comment (uriyyo) [reply pydantic/main.py]: Makes sense. I have created function `update_model_forward_refs` at `typing.py`.

File: pydantic/main.py (hunk 2 above)
Comment (uriyyo) [reply pydantic/main.py]: Fixed

**Comment by uriyyo** [comment]
@samuelcolvin I have updated PR. Could you please review it again?

File: changes/2588-uriyyo.md [hunk 4]
```diff
@@ -0,0 +1 @@
+Try to evaluate forward refs after model created.
```
Comment (PrettyWood) [inline changes/2588-uriyyo.md:1]:
```suggestion
Try to evaluate forward refs automatically at model creation.
```

File: docs/usage/postponed_annotations.md [hunk 5]
```diff
@@ -45,9 +45,8 @@ Resolving this is beyond the call for *pydantic*: either remove the future impor
 
 ## Self-referencing Models
 
-Data structures with self-referencing models are also supported, provided the function
-`update_forward_refs()` is called once the model is created (you will be reminded
-with a friendly error message if you forget).
+Data structures with self-referencing models are also supported. Self-referencing fields will be automatically
+resolved after was created.
```
Comment (PrettyWood) [inline docs/usage/postponed_annotations.md:49]:
```suggestion
resolved after model creation.
```

File: pydantic/main.py [hunk 6]
```diff
@@ -374,6 +373,8 @@ def is_untouched(v: Any) -> bool:
         cls = super().__new__(mcs, name, bases, new_namespace, **kwargs)
         # set __signature__ attr only for model class, but not for its instances
         cls.__signature__ = ClassAttribute('__signature__', generate_model_signature(cls.__init__, fields, config))
+        cls.__try_update_forward_refs__()
```
Comment (PrettyWood) [inline pydantic/main.py:376]: What is the impact on perf for models without forward refs? Do we have a way to know whether or not we should try to update?

**Review by PrettyWood — COMMENTED** [review]
Cool feature! Nice to be able to do everything automatically

**Comment by PrettyWood** [comment]
Hi @uriyyo 
Can you please rebase/merge and update? 🙏

File: pydantic/main.py (hunk 6 above)
Comment (uriyyo) [reply pydantic/main.py:376]: I wrote a simple benchmark:
```py
import timeit
from pydantic import BaseModel


def callback():
    class Order(BaseModel):
        a: int
        b: float
        c: str


t = timeit.timeit(
    callback,
    number=10_000,
    globals={"BaseModel": BaseModel},
)

print(f'{t:.5f}')
```

Results:
* With `__try_update_forward_refs__` - 3.17182
* Without `__try_update_forward_refs__` - 3.15772

**Review by PrettyWood — COMMENTED** [review]
Yes I run the benchmark seems ok performance-wise

### This branch
```
▶ BENCHMARK_REPEATS=20 make benchmark-pydantic
python benchmarks/run.py pydantic-only
testing pydantic, 20 times each
    pydantic ( 1/20) time=0.622s, success=48.65%
    pydantic ( 2/20) time=0.617s, success=48.65%
    pydantic ( 3/20) time=0.598s, success=48.65%
    pydantic ( 4/20) time=0.596s, success=48.65%
    pydantic ( 5/20) time=0.597s, success=48.65%
    pydantic ( 6/20) time=0.606s, success=48.65%
    pydantic ( 7/20) time=0.595s, success=48.65%
    pydantic ( 8/20) time=0.598s, success=48.65%
    pydantic ( 9/20) time=0.591s, success=48.65%
    pydantic (10/20) time=0.597s, success=48.65%
    pydantic (11/20) time=0.595s, success=48.65%
    pydantic (12/20) time=0.605s, success=48.65%
    pydantic (13/20) time=0.603s, success=48.65%
    pydantic (14/20) time=0.601s, success=48.65%
    pydantic (15/20) time=0.611s, success=48.65%
    pydantic (16/20) time=0.603s, success=48.65%
    pydantic (17/20) time=0.600s, success=48.65%
    pydantic (18/20) time=0.596s, success=48.65%
    pydantic (19/20) time=0.592s, success=48.65%
    pydantic (20/20) time=0.605s, success=48.65%
    pydantic best=0.591s, avg=0.601s, stdev=0.008s

    pydantic best=98.503μs/iter avg=100.240μs/iter stdev=1.335μs/iter version=1.8.2
```

### 1.8
```
▶ BENCHMARK_REPEATS=20 make benchmark-pydantic
python benchmarks/run.py pydantic-only
testing pydantic, 20 times each
    pydantic ( 1/20) time=0.608s, success=48.65%
    pydantic ( 2/20) time=0.588s, success=48.65%
    pydantic ( 3/20) time=0.616s, success=48.65%
    pydantic ( 4/20) time=0.632s, success=48.65%
    pydantic ( 5/20) time=0.599s, success=48.65%
    pydantic ( 6/20) time=0.588s, success=48.65%
    pydantic ( 7/20) time=0.591s, success=48.65%
    pydantic ( 8/20) time=0.595s, success=48.65%
    pydantic ( 9/20) time=0.589s, success=48.65%
    pydantic (10/20) time=0.612s, success=48.65%
    pydantic (11/20) time=0.599s, success=48.65%
    pydantic (12/20) time=0.594s, success=48.65%
    pydantic (13/20) time=0.623s, success=48.65%
    pydantic (14/20) time=0.591s, success=48.65%
    pydantic (15/20) time=0.592s, success=48.65%
    pydantic (16/20) time=0.595s, success=48.65%
    pydantic (17/20) time=0.595s, success=48.65%
    pydantic (18/20) time=0.600s, success=48.65%
    pydantic (19/20) time=0.625s, success=48.65%
    pydantic (20/20) time=0.587s, success=48.65%
    pydantic best=0.587s, avg=0.601s, stdev=0.014s

    pydantic best=97.825μs/iter avg=100.158μs/iter stdev=2.254μs/iter version=1.8
```

**Review by PrettyWood — APPROVED** [review]
LGTM!

**Comment by samuelcolvin** [comment]
this is great, thank you so much.

**Pull request merged**